src/werkzeug/sansio/multipart.py

```diff
@@ -0,0 +1,251 @@
+import re
+from dataclasses import dataclass
+from enum import auto
+from enum import Enum
+from typing import cast
+from typing import List
+from typing import Optional
+from typing import Tuple
+from typing import Union
+
+from .._internal import _to_bytes
+from .._internal import _to_str
+from ..datastructures import Headers
+from ..exceptions import RequestEntityTooLarge
+from ..http import parse_options_header
+
+
+@dataclass
+class Preamble:
+    data: bytes
+
+
+@dataclass
+class Field:
+    name: str
+    headers: Headers
+
+
+@dataclass
+class File:
+    name: str
+    filename: str
+    headers: Headers
+
+
+@dataclass
+class Data:
+    data: bytes
+    more_data: bool
+
+
+@dataclass
+class Epilogue:
+    data: bytes
+
+
+@dataclass
+class NeedData:
+    pass
+
+
+Event = Union[Preamble, Field, File, Data, Epilogue, NeedData]
+
+
+class State(Enum):
+    PREAMBLE = auto()
+    PART = auto()
+    DATA = auto()
+    EPILOGUE = auto()
+    COMPLETE = auto()
+
+
+# Multipart line breaks MUST be CRLF (\r\n) by RFC-7578, except that
+# many implementations break this and either use CR or LF alone.
+LINE_BREAK = b"(?:\r\n|\n|\r)"
+BLANK_LINE_RE = re.compile(b"(?:\r\n\r\n|\r\r|\n\n)", re.MULTILINE)
+LINE_BREAK_RE = re.compile(LINE_BREAK, re.MULTILINE)
+# Header values can be continued via a space or tab after the linebreak, as
+# per RFC2231
+HEADER_CONTINUATION_RE = re.compile(b"%s[ \t]" % LINE_BREAK, re.MULTILINE)
+
+
+class MultipartDecoder:
+    """Decodes a multipart message as bytes into Python events.
+
+    The part data is returned as available to allow the caller to save
+    the data from memory to disk, if desired.
+    """
+
+    def __init__(
+        self,
+        boundary: bytes,
+        max_form_memory_size: Optional[int] = None,
+    ) -> None:
+        self.buffer = bytearray()
+        self.complete = False
+        self.max_form_memory_size = max_form_memory_size
```

Comment:
I'd prefer us to fix #1513 in a separate change, rather than with this (will make this easier for me to manage and hopefully merge). Maybe with the body attribute idea I mentioned.